debezium-core/src/main/java/io/debezium/pipeline/spi/ChangeRecordEmitter.java

```diff
@@ -30,6 +31,20 @@
     OffsetContext getOffset();
 
     public interface Receiver {
-        void changeRecord(DataCollectionSchema schema, Operation operation, Object key, Struct value, OffsetContext offset) throws InterruptedException;
+        void changeRecord(DataCollectionSchema schema,
+                          Operation operation,
+                          Object key, Struct value,
+                          OffsetContext offset)
+                throws InterruptedException;
+
+        default void changeRecord(DataCollectionSchema schema,
+                                  Operation operation,
+                                  Object key, Struct value,
+                                  OffsetContext offset,
+                                  ConnectHeaders headers)
+                throws InterruptedException {
+            throw new RuntimeException("Not implemented yet in " + this.getClass().getName());
```

Comment:
Is this going to be addressed as part of this PR?

Reply:
I think it shouldn't. What do you suggest?
From my point it's just a helper when you create a new Connector implementation and you use a ChangeRecordEmitter that's not capable of creating headers (that doesn't implement the new signature of changeRecord(...), then you get warned.
Other alternative would be a extended interface to ChangeRecordEmitter, like ChangeRecordWithHeadersEmitter.

Reply:
I'd prefer this to be a plain interface method, i.e. non-defaulted. `Receiver` is a callback interface that's implemented by Debezium core, not connectors. Hence there shouldn't be a case where implementations don't support headers.

Reply:
But ChangeRecordEmitter is implemented by Connectors and they are the callers for receiver which is the nested interface.

Shouldn't it be optional for them to choose if they send with headers or without?
Then we have 2 options:
- like I did now: one method with two signatures in one interface, where the one with headers is optional (comes with default implementation), the default implementation in debezium-core offers both, but Connectors can implement their own ChangeRecordEmitter with their own Receiver and freely choose if they send with or without headers (they just need to make sure in their Reiceiver they provide an implementation with headers)
- two interfaces `Receiver` and `ReceiverWithHeaders`: currently Receiver is only implemented in debezium-core, but everyone could provide it's own receiver because its not closed (open-close-principle). I think we can't assume everyone is using our base implementation in debezium-core, that's why I think this option might be a bad choice. also it comes with many more code changes, the default implementation in debezium-core gets complex as it needs implementation for `void emitChangeRecords(DataCollectionSchema schema, Receiver receiver) throws InterruptedException;` and for `void emitChangeRecords(DataCollectionSchema schema, ReceiverWithHeaders receiver) throws InterruptedException;`.

If you want no one to implement Receiver interface we should extract it from ChangeRecordEmitter and make private/package-private.

Reply:
`Receiver` is not meant to be implemented by connectors, only by core. About `ChangeRecordEmitter`, changing that could be considered breaking, but so far we didn't consider any APIs/SPIs geared towards connector implementors as public contracts. I.e. we don't (yet) pursue a model with a stable API to be used by connector implementors unbeknownst to us.

I'd simply change the `Receiver` method (adding the headers param) and adjust all existing emitter implementations. Either passing the headers, or null if they can't. Let's not overcomplicate things :)

Reply:
I'll remove the default implementation, it's not really required any longer.